examples/interpreter/physl.cpp

```diff
@@ -173,43 +219,112 @@ int main(int argc, char* argv[])
+        ast = load_ast(ast_dump_file);
+        code_source_name = fs::path(ast_dump_file).filename().string();
     }
+    // Read PhySL source code from a file or the provided argument
     else
     {
-        // Interpret first argument as the file name for the PhySL code
-        user_code = read_user_code(positional_args[0]);
-        code_source_name = fs::path(positional_args[0]).filename().string();
-        first_index = 1;
+        if (vm.count("code") != 0)
+        {
+            // Execute code as given directly on the command line
+            user_code = vm["code"].as<std::string>();
+            code_source_name = "<command_line>";
+        }
+        else if (positional_args.size() > 0)
+        {
+            // Interpret first argument as the file name for the PhySL code
+            user_code = read_user_code(positional_args[0]);
+            code_source_name = fs::path(positional_args[0]).filename().string();
+            // The rest of the positional arguments are arguments for the script
+            first_index = 1;
+            code_is_file = true;
+        }
+        else
+        {
+            std::cout << "No code was provided.\n";
+
+            return -1;
+        }
+
+        try
+        {
+            // Compile the given code into AST
+            ast = phylanx::ast::generate_ast(user_code);
+
+            // Apply transformation rules to AST, if requested
+            if (vm.count("transform") != 0)
+            {
+                std::string const transform_rules =
+                    read_user_code(vm["transform"].as<std::string>());
+
+                ast = phylanx::ast::transform_ast(
+                    ast, phylanx::ast::generate_transform_rules(transform_rules));
+            }
+
+            // Dump the AST to a file, if requested
+            if (vm.count("dump-ast") != 0)
+            {
+                std::string dump_file = vm["dump-ast"].as<std::string>();
+                // If no dump file is specified but PhySL code is read from a
+                // file then use the file name with .ast extension
+                if (dump_file == "<none>")
+                {
+                    if (!code_is_file)
+                    {
+                        std::cerr << "physl: command line handling: exception "
+                                     "caught: the required path argument for "
+                                     "option '--dump-ast' is missing\n";
+                        return -1;
+                    }
+                    dump_file = fs::path(positional_args[0])
+                                    .replace_extension("ast")
+                                    .string();
+                }
+                dump_ast(ast, dump_file);
+            }
+        }
+        catch (std::exception const& e)
+        {
+            std::cout << "physl: exception caught:\n" << e.what() << "\n";
+            return -1;
+        }
```

Comment:
The interpreter can apply an AST transformation. Would it be possible to move the writing of the generated AST further down to enable writing the transformed AST instead of the originally generated one?